### ISSUE TITLE
Does current pulsar implementation support differentiable opacity optimization

### ISSUE THREAD

**Issue by Wang-Tianyu** [facebookresearch/pytorch3d#501]
I am playing around with the pulsar example 
[https://github.com/facebookresearch/pytorch3d/blob/master/docs/examples/pulsar_multiview.py](url)

I notice that the vertex opacity parameter vert_opy is initialized with requires_grad=True but never used.
Thus I added vert_opy as input to the renderer and add vert_opy into the optimizer parameter list.
However, the value of vertex opacity never changes during the optimization process.

In the current implementation, will opacity receive gradients?
If they do receive gradients, is it possible to provide an example with opacity used?

**Comment by classner** [comment]
Hi @Wang-Tianyu !

Good catch! I set it up to be used as a parameter and in the end did not use it in the optimization. It is absolutely supported! Make sure to add it to both, renderer.forward as keyword argument `opacity` in BxN shape (which it sounds like you did) and adjust its learning rate accordingly. You probably also have to add clamping to its values, because it must remain in [0, 1] and adding gradients to it does not necessarily do that (l. 187 does this for the vert_col property).

There's currently no open source example for using it, but we use it for the NeRF-scene optimizations in the tech report (https://arxiv.org/abs/2004.07484).

**Comment by Wang-Tianyu** [comment]
Hi, classner,

    Thanks for you quick reply.
    I checked my code again and I did add opacity = self.vert_opy.expand(n_views, -1) at around line 112.
    But when I try to print(model.vert_opy,grad) right after loss.backward() around line 183, I still get None.
    vert_pos, vert_col and vert_rad all have grad value.
    I can confirm that opacity indeed influences the rendering process since when I set it to 0.1, spheres are not quite visible.

    Did I miss anything?
   
    Also, below is the running results of the multiview optimization script. I wait until the loss does not drop anymore.

![multiview](https://user-images.githubusercontent.com/16913207/103615289-6ee66a80-4f7e-11eb-92d8-a680db342214.gif)

    I notice that there are still some points floating around. Is this the correct output or did I make any mistakes?
    After trying different scenes (generated with the create_multiview script), it seems that spheres located at the center of scenes can be better reconstructed compared to spheres located at scene boundaries. I cannot figure out why this is the case.


Cheers

**Comment by Wang-Tianyu** [comment]
I realized that the provided script uses L2 loss while in the tech report all results are obtained with L1 loss.
After changing L2 loss to L1 loss, the results appear to be better.

![multiview](https://user-images.githubusercontent.com/16913207/104089082-f3f3bb80-52bf-11eb-9230-43dc5f885b37.gif)

But objects that are away from the center of the scene are still not very well reconstructed compared with objects that are located close to the center.

I still cannot get any gradient for opacity.

**Comment by classner** [comment]
@Wang-Tianyu fix is under way and will land soon in master! Sorry for the delay!

Cause was a simple mis-routing of gradient calculation flags in the Python interface. This effectively was always 'false' for opacity in this version and left the gradients zeroed out.

**Comment by xuelin-chen** [comment]
Hi @classner,

Thanks for this great Pulsar renderer! It is indeed so easy to use. ;-)

But, I was wondering if the opacity gradient issue mentioned in this thread has been solved in the latest version?
Also is there any plan to open-source the code for the experiments, like particularly the code to implement the NeRF-scene optimization where opacity optimization is included, in the Pulsar paper?:-)

**Comment by classner** [comment]
Hi @xuelin-chen !

Great that you like the renderer! Indeed, the opacity gradient issue has already long been solved! There are currently no plans on open-sourcing this code, unfortunately. It is also important to note that we do not recommend this strategy for general-purpose 3D reconstruction - this was mainly an experiment to demonstrate the effectiveness of the gradients.

**Comment by xuelin-chen** [comment]
I see, thanks for your reply!
